examples/mario/src/main/scala/Effects.scala

```diff
@@ -0,0 +1,90 @@
+package mario
+
+import org.scalajs.dom
+import org.scalajs.dom.{KeyboardEvent, document}
+import org.scalajs.dom.raw.{Event, HTMLAudioElement, TouchEvent}
+import scalm.{Cmd, Sub, Task}
+import scalm.Sub.ofTotalObservable
+
+object Effects {
+
+  val requestAnimationFrameSub: Sub[Double] = ofTotalObservable[Double](
+    "requestAnimation", { observer =>
+      var handle = 0
+      def loop(time: Double): Unit = {
+        observer.onNext(time)
+        handle = dom.window.requestAnimationFrame(loop)
+      }
+      handle = dom.window.requestAnimationFrame(loop)
+      () =>
+        dom.window.cancelAnimationFrame(handle)
+    }
+  )
+
+  def keyPressSub[M](keyCode: Int, msg: M): Sub[M] = ofTotalObservable[M](
+    s"keyDown$keyCode", { observer =>
+      val listener = (keyEvent: KeyboardEvent) => {
+        if (keyEvent.keyCode == keyCode) observer.onNext(msg)
+      }
+      dom.window.addEventListener("keydown", listener)
+      () =>
+        dom.window.removeEventListener("keydown", listener)
```

Comment:
I think there is a trap here: the type of `listener` is `scala.Function1[KeyboardEvent, Unit]`. When you call `addEventListener`, `listener` is implicitly converted to `scalajs.js.Function1[KeyboardEvent, Unit]`, and you never get the underlying reference. Consequently, when you call `removeEventListener` you don’t pass a correct reference.

The solution is either to get a `js.Function1` from the beginning:

~~~ scala
val listener: js.Function1[KeyboardEvent, Unit] = …
~~~

Or, to use `util.Functions.fun`, which does the same but requires you to write less types:

~~~ scala
val listener = fun { (kbd: KeyboardEvent) => … }
~~~

Reply:
Oh! It totally makes sense.